src/Workleap.Extensions.OpenApi.Analyzers/CompareTypedResultWithAnnotationAnalyzer.cs

```diff
@@ -0,0 +1,279 @@
+            // Will be Supported in .NET 9
+            Add("Microsoft.AspNetCore.Http.HttpResults.InternalServerError", 500);
+            Add("Microsoft.AspNetCore.Http.HttpResults.InternalServerError`1", 500);
+            // Workleap's definition of the InternalServerError type result for other .NET versions
+            Add("Workleap.Extensions.OpenAPI.TypedResult.InternalServerError", 500);
+            Add("Workleap.Extensions.OpenAPI.TypedResult.InternalServerError`1", 500);
+
+            return dictionary;
+
+            void Add(string metadata, int statusCode)
+            {
+                var type = compilation.GetTypeByMetadataName(metadata);
+                if (type is not null)
+                {
+                    dictionary.Add(type, statusCode);
+                }
+            }
+        }
+
+        private static Dictionary<int, ITypeSymbol> InitializeStatusCodeMapHttpResultMap(Compilation compilation)
+        {
+            var dictionary = new Dictionary<int, ITypeSymbol>();
+            Add(200, "Microsoft.AspNetCore.Http.HttpResults.Ok");
+            Add(201, "Microsoft.AspNetCore.Http.HttpResults.Created");
+            Add(202, "Microsoft.AspNetCore.Http.HttpResults.Accepted");
+            Add(204, "Microsoft.AspNetCore.Http.HttpResults.NoContent");
+            Add(400, "Microsoft.AspNetCore.Http.HttpResults.BadRequest");
+            Add(401, "Microsoft.AspNetCore.Http.HttpResults.UnauthorizedHttpResult");
+            Add(404, "Microsoft.AspNetCore.Http.HttpResults.NotFound");
+            Add(409, "Microsoft.AspNetCore.Http.HttpResults.Conflict");
+            Add(422, "Microsoft.AspNetCore.Http.HttpResults.UnprocessableEntity");
+            Add(500, "Microsoft.AspNetCore.Http.HttpResults.InternalServerError");
+            Add(500, "Workleap.Extensions.OpenAPI.TypedResult.InternalServerError");
+
+            return dictionary;
+
+            void Add(int statusCode, string metadata)
+            {
+                var type = compilation.GetTypeByMetadataName(metadata);
+                if (type is not null)
+                {
+                    dictionary.Add(statusCode, type);
+                }
+            }
+        }
+
+        public void AnalyzeClassDeclaration(SymbolAnalysisContext context)
+        {
+            var methodSymbol = (IMethodSymbol)context.Symbol;
+            var returnType = methodSymbol.ReturnType;
+
+            if (returnType is INamedTypeSymbol typedReturnType)
+            {
+                var iResultTypeSymbol = context.Compilation.GetTypeByMetadataName("Microsoft.AspNetCore.Http.IResult");
+                if (iResultTypeSymbol is null)
+                {
+                    return;
+                }
+
+                // Check if the return type is of Task<IResult> or Task<Result<>>. If yes, then keep the inner type.
+                if (SymbolEqualityComparer.Default.Equals(typedReturnType.ConstructedFrom, this.TaskOfTSymbol))
```

Comment:
Should it support ValueTask?

Reply:
Also, you may want to create a new method like `UnwrapTask` to make it more explicit?

Reply:
good point, will extract into tis own method and do both checks